pkg/autoscaler/statforwarder/forwarder.go

```diff
@@ -0,0 +1,253 @@
+//    Lease -> IP mapping.
+// 2. Creates/updates the corresponding K8S Service and Endpoints.
+// 3. Can be used to forward the metrics falling in a bucket based on
+//    the holder IP. (This is a TODO)
+type Forwarder struct {
+	selfIP          string
+	logger          *zap.SugaredLogger
+	kc              kubernetes.Interface
+	serviceLister   corev1listers.ServiceLister
+	endpointsLister corev1listers.EndpointsLister
+	// bs is the BucketSet including all Autoscaler buckets.
+	bs *hash.BucketSet
+	// Lock for leaseHolders.
+	leaseLock sync.RWMutex
+	// leaseHolders stores the Lease -> IP relationships.
+	leaseHolders map[string]string
+}
+
+// New creates a new Forwarder.
+func New(ctx context.Context, logger *zap.SugaredLogger, kc kubernetes.Interface, selfIP string, bs *hash.BucketSet) *Forwarder {
+	ns := system.Namespace()
+	bkts := bs.Buckets()
+	endpointsInformer := endpointsinformer.Get(ctx)
+	f := Forwarder{
+		selfIP:          selfIP,
+		logger:          logger,
+		kc:              kc,
+		serviceLister:   serviceinformer.Get(ctx).Lister(),
+		endpointsLister: endpointsInformer.Lister(),
+		bs:              bs,
+		leaseHolders:    make(map[string]string, len(bkts)),
+	}
+
+	leaseInformer := leaseinformer.Get(ctx)
+	leaseInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: reconciler.NamespaceFilterFunc(ns),
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    f.leaseUpdated,
+			UpdateFunc: controller.PassNew(f.leaseUpdated),
+			// TODO(yanweiguo): Set up DeleteFunc.
+		},
+	})
+
+	return &f
+}
+
+func (f *Forwarder) leaseUpdated(obj interface{}) {
+	l := obj.(*coordinationv1.Lease)
+	ns, n := l.Namespace, l.Name
+
+	if !f.bs.HasBucket(n) {
+		// Skip this Lease as it is not for Autoscaler.
+		return
+	}
+
+	if l.Spec.HolderIdentity == nil || *l.Spec.HolderIdentity == "" {
+		f.logger.Debugf("HolderIdentity not found for Lease %s/%s", ns, n)
+		return
+	}
+
+	identity := *l.Spec.HolderIdentity
+	if h, ok := f.getHolder(n); ok && h == identity {
+		// Already up-to-date.
+		return
+	}
+
+	f.setHolder(n, identity)
+
+	if identity != f.selfIP {
+		// Skip creating/updating Service and Endpoints as not the holder.
+		return
+	}
+
+	if err := f.createService(ns, n); err != nil {
+		f.logger.Errorf("Failed to create Service for Lease %s/%s: %v", ns, n, err)
+		return
+	}
+	f.logger.Infof("Created Service for Lease %s/%s", ns, n)
+
+	if err := f.createOrUpdateEndpoints(ns, n); err != nil {
+		f.logger.Errorf("Failed to create Endpoints for Lease %s/%s: %v", ns, n, err)
+		return
+	}
+	f.logger.Infof("Created Endpoints for Lease %s/%s", ns, n)
+}
+
+func (f *Forwarder) checkService(ns, n string) error {
+	return retry.OnError(retry.DefaultRetry, exceptNotFound, func() error {
+		_, err := f.serviceLister.Services(ns).Get(n)
+		return err
+	})
+}
+
+func (f *Forwarder) createService(ns, n string) error {
+	err := f.checkService(ns, n)
```

Comment:
Why do we need to check service before creating it? Can we try to create it, and if api returns `IsAlreadyExists` error then we stop retrying?

Reply:
Good point.

Hmm, I was thinking about the pattern that

 1. Get
 2. Create if not found.